### PR TITLE
Remove tier object from memtier API

### PR DESCRIPTION
**Pull request by michalbiesek**
I did not modify naming in external headers, feel free to give feedback there as well.

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/615)

### CODE REVIEW

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 3 of 5 files at r2, 2 of 3 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/615)* status: all files reviewed, 2 unresolved discussions (waiting on @jschmieg, @michalbiesek, and @PatKamin)

___
*[src/memkind_memtier.c, line 100 at r4](https://reviewable.io/reviews/memkind/memkind/615#-MXfIC_d43PwPA4lCv-x:-MXfIC_d43PwPA4lCv-y:b-bob1b5) ([raw file](https://github.com/memkind/memkind/blob/5888a9685021925a743852ca3839bc37b4dca931/src/memkind_memtier.c#L100)):*
> ```C
>     for (i = 0; i < builder->size; ++i) {
>         if (kind == builder->cfg[i].kind) {
>             log_err("Kind is already in builder.");
> ```

This is confusing. API call is named "..add_tier" and here is "kind". Maybe rename API call to *_create_tier_form_kind ?
___
*[src/memkind_memtier.c, line 192 at r4](https://reviewable.io/reviews/memkind/memkind/615#-MXfJeaQ3WhMvw-1F83Q:-MXfJeaQ3WhMvw-1F83R:b-muutnf) ([raw file](https://github.com/memkind/memkind/blob/5888a9685021925a743852ca3839bc37b4dca931/src/memkind_memtier.c#L192)):*
> ```C
> }
> 
> MEMKIND_EXPORT void *memtier_kind_malloc(struct memtier_kind *kind, size_t size)
> ```

Again, the API is simpler in number of calls, but it is more confusing to me. I agree with your changes and they really simplifies code but I think we should change function names. Maybe '"memtier_malloc" will be better here?

File: man/memkind_memtier.3 [hunk 1]
```diff
@@ -16,15 +16,11 @@ functionality is considered as stable API (STANDARD API).
 .sp
 .B "TIER MANAGEMENT:"
 .sp
-.BI "struct memtier_tier *memtier_tier_new(memkind_t " "kind" );
```
Comment (PatKamin) [inline man/memkind_memtier.3:19]: Remove this as well https://github.com/memkind/memkind/blob/feat-memory-tiering/man/memtier.7#L17

File: src/memkind_memtier.c [hunk 2]
```diff
@@ -90,10 +90,17 @@ MEMKIND_EXPORT void memtier_builder_delete(struct memtier_builder *builder)
 MEMKIND_EXPORT int memtier_builder_add_tier(struct memtier_builder *builder,
                                             memkind_t kind, unsigned kind_ratio)
 {
+    int i;
```
Comment (PatKamin) [inline src/memkind_memtier.c:93]: nit: Add an empty line below.

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 6 of 6 files at r5, 5 of 5 files at r6, 3 of 3 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/615)* status: all files reviewed, 3 unresolved discussions (waiting on @jschmieg and @michalbiesek)

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/615)* status: 0 of 9 files reviewed, 3 unresolved discussions (waiting on @bratpiorka and @jschmieg)

___
*[include/memkind_memtier.h, line 108 at r8](https://reviewable.io/reviews/memkind/memkind/615#-MXfEQUAKRgsuWCsShG5-r8-108:-MXg_1zV2e4a4uu7sijw:b76mk7l) ([raw file](https://github.com/memkind/memkind/blob/38ee0fdb97805ca9d1db97f398b2a459257b363b/include/memkind_memtier.h#L108)):*
<details><summary><i>Previously, jschmieg wrote…</i></summary><blockquote>

IMHO: Allocating from specific "memkind_kind" should be called: memtier_kind_malloc
Allocating from tiered memory should be called: memtier_malloc
and memtier_builder_construct_kind should create something like "memtier_memory"
</blockquote></details>

Done.

memtier tier is now memtier_memory
allocate to memtier_memory is done by memtier_*alloc API
allocate to memkind kind is done by memtier_kind*alloc API
___
*[src/memkind_memtier.c, line 100 at r4](https://reviewable.io/reviews/memkind/memkind/615#-MXfIC_d43PwPA4lCv-x:-MXg_Hii0J9cQPRByvrn:b-kzxyab) ([raw file](https://github.com/memkind/memkind/blob/5888a9685021925a743852ca3839bc37b4dca931/src/memkind_memtier.c#L100)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

This is confusing. API call is named "..add_tier" and here is "kind". Maybe rename API call to *_create_tier_form_kind ?
</blockquote></details>

I prefer add_tier since I don't get any "created" object back
___
*[src/memkind_memtier.c, line 192 at r4](https://reviewable.io/reviews/memkind/memkind/615#-MXfJeaQ3WhMvw-1F83Q:-MXg7_NC4K3wqlxSu_l6:b1maxlj) ([raw file](https://github.com/memkind/memkind/blob/5888a9685021925a743852ca3839bc37b4dca931/src/memkind_memtier.c#L192)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

Again, the API is simpler in number of calls, but it is more confusing to me. I agree with your changes and they really simplifies code but I think we should change function names. Maybe '"memtier_malloc" will be better here?
</blockquote></details>

Done - I provide changes aligned with @jschmieg suggestion.
___
*[src/memkind_memtier.c, line 93 at r8](https://reviewable.io/reviews/memkind/memkind/615#-MXfEQUAKRgsuWCsShG4-r8-93:-MXg7X9o3el3y2PrrZHi:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/38ee0fdb97805ca9d1db97f398b2a459257b363b/src/memkind_memtier.c#L93)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

nit: Add an empty line below.
</blockquote></details>

Done.

**Review by bratpiorka — COMMENTED** [review]
:lgtm:

> Reviewed 4 of 6 files at r9, 4 of 6 files at r10, 2 of 3 files at r11, 5 of 7 files at r13, 5 of 5 files at r14, 4 of 4 files at r15.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/615#-:-MXgd1cc-MaGtPmYYejK:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @jschmieg)

**Review by jschmieg — APPROVED** [review]

**Review by PatKamin — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 6 files at r10, 1 of 3 files at r11, 2 of 7 files at r13, 1 of 5 files at r14, 4 of 4 files at r15.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/615#-:-MXk6F5PD45NEFRFtEoG:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @jschmieg)

**Pull request merged**